### DISCUSSION TITLE
Jolt and Physics Questions

### DISCUSSION THREAD

**Discussion by MCVenor** [godot-jolt/godot-jolt#901] [Questions] [answered]
Hi,

i am using the Godot Jolt extension, latests [0.13.0-beta1](https://github.com/godot-jolt/godot-jolt/releases/tag/v0.13.0-beta1) Pre-release
With Godot 4.3 (starting with [4.3-beta1]
(Up until yesterday i have been using an older Version, i realy enjoy the Godot Jolt Extension! Great Work)

Now i have a couple questions that i tried to find awnsers to but couldnt find.
I have looked at Godots original move_and_collide. It returns a result that can have multiple Collision packed in a KinematicCollision3D.
But since i am using Godot Jolt, i tried to check out where the call to move_and_collide leads me. But i couldnt find the exact glue/binding point. Thou i did find the explanation with the 3 Steps. (depenetration, cast and finally adjusted movement).

Before i list the questions that i have, here is the scenario i tried to test.
For testing, i made 16 StaticBody3Ds and let MoveAndCollided() fall a CharacterBody3D onto them. The result was 1 single collision.

I tested it with:

    16 StaticBody3D EACH having a CollisionShape3D with BoxShape3D
    1 StaticBody3D having 16 CollisionShape3D with BoxShape3D
    1 StaticBody3D having 16 StaticBody3D EACH having a CollisionShape3D with BoxShape3D

        KinematicCollision3D bel = MoveAndCollide(Velocity*0.1f);

        if (bel != null)
        {
            int i = bel.GetCollisionCount();
            C.Log(i.ToString());
        }

Not a single time did i get a CollisionCount greater than 1.
So i moved the CharacterBody3D with its CollisionShape **INTO** the 16 boxes (its offcenter), like so:
![e3e3718727fe7488845cdb2b2563d52705173b91](https://github.com/godot-jolt/godot-jolt/assets/12583072/e23e0dc0-adcd-473f-919d-ef187c504b8c)
Blue are the 16 StaticBody3D , the light white color is the CharacterBody3D's CollisionShape3D BoxShape
Same result, only 1 collision was reported from GetCollisionCount().

That said here are the Questions:

- What am i missing, under what scenario will i get more then 1 from GetCollisionCount() returned?
- Where is the Glue/Binding point from Godots Original move_and_collide or other methods to Godot Jolts Methods (or just Jolt methods) ? (this way i could at least try to understand it by looking it up)

General Questions (i tried to check the Jolt Manual, but again, i couldnt find definitive awnsers)
- i assume a RayCast is greatly more performant than a ShapeCast give the same distance, is this right?
(my assumtion is based on, that one could loop throu all vertices of a Shape OR throu its mathematical Body (like sphere) on every step as opposed to a ray, who doesnt have any Shape or Body)
- for RayCast and ShapeCast, how many "Steps" are done? For example; if i cast with Vector(0,1,0) does the physics check every 0.01 or 0.001 units?
- Is there a special number/case where ShapeCast beats RayCasting or the other way around?
(say i should stop using 4 rays down and instead ShapeCast a Box downwards to floorChecking
OR should i stop using the ShapeCast Box and instead just use 4 more prependicular rays to make a "box" bottom myself)

**Comment by mihe** [comment, marked as answer]
> But since i am using Godot Jolt, i tried to check out where the call to move_and_collide leads me. But i couldnt find the exact glue/binding point.

`PhysicsBody3D.move_and_collide` is essentially a very thin wrapper around `PhysicsServer3D.body_test_motion`, the implementation for which you'll find [here](https://github.com/godot-jolt/godot-jolt/blob/6ffe4ff2b681143fd38565fd3e416b1a42bf6f2d/src/spaces/jolt_physics_direct_space_state_3d.cpp#L676-L775).

> Blue are the 16 StaticBody3D , the light white color is the CharacterBody3D's CollisionShape3D BoxShape
> Same result, only 1 collision was reported from GetCollisionCount().

I'm struggling a bit to understand your exact scenario. If you're able to provide a minimal repro project instead that would be preferable.

> i assume a RayCast is greatly more performant than a ShapeCast give the same distance, is this right?

Without having looked at the exact numbers, almost definitely, yes.

> for RayCast and ShapeCast, how many "Steps" are done? For example; if i cast with Vector(0,1,0) does the physics check every 0.01 or 0.001 units?

I haven't looked over Jolt's exact implementation for every type of shape, but ray-casts are typically done using a simple ray intersection test, so they're not done incrementally. Things presumably get a bit more complicated when you're dealing with things like compound shapes and polygonal shapes, but at its core it's still just a ray intersection test.

Shape-casts are however done using a binary search, which is not how Jolt typically does it, but is sort of forced upon any physics implementation in Godot due to how the `cast_motion`/`body_test_motion` interface is designed. Godot Physics uses a hardcoded number of collision checks, which is set to 8, whereas this extension calculates the amount of steps based on the distance. You can see the formula for it [here](https://github.com/godot-jolt/godot-jolt/blob/6ffe4ff2b681143fd38565fd3e416b1a42bf6f2d/src/spaces/jolt_physics_direct_space_state_3d.cpp#L848-L850).

> Is there a special number/case where ShapeCast beats RayCasting or the other way around?

I can't put a number on it, unfortunately, but ray-casts are for sure cheaper in pretty much every case. I would imagine you could probably do dozens of ray-casts in the same time it takes to do a single shape-cast. It also matters greatly what type of shape you're shape-casting of course, but I would suggest you do your own profiling for a definitive answer.

**Comment by MCVenor** [comment]
This was precicly what i was looking for. Ty.
I will soonish provide a MRP.

**Comment by MCVenor** [comment]
Yes i came across that "8", along with [some 7y old comment](https://github.com/godotengine/godot/blob/b75f0485ba15951b87f1d9a2d8dd0fcd55e178e4/servers/physics_3d/godot_space_3d.cpp#L650) 

----

bool GodotSpace3D::test_body_motion(GodotBody3D* p_body, const PhysicsServer3D::MotionParameters &p_parameters, PhysicsServer3D::MotionResult* r_result) 

  //give me back regular physics engine logic
  //this is madness
  //and most people using this function will think
  //what it does is simpler than using physics
  //this took about a week to get right..
  //but is it right? who knows at this point..

----

Anyway here is the MRP, i hope i did it right.
I cant upload it here, so i uploaded it here https://file.io/28TS3Wab4psX .
should i upload it somewhere else, let me know where.

Inside the Project are 3 Scenes test_1, test_2 and test_3.
They all are very similar, the only difference is that composition of the little platform.
In none of those examples i am able to get a [KinematicCollision3D.GetCollisionCount()](https://docs.godotengine.org/en/stable/classes/class_kinematiccollision3d.html#class-kinematiccollision3d-method-get-collision-count) result of more than 1. 
(witch you can find in the script of the character controller)

Also thank you for the Link 
_PhysicsBody3D.move_and_collide is essentially a very thin wrapper around PhysicsServer3D.body_test_motion, the implementation for which you'll find [here](https://github.com/godot-jolt/godot-jolt/blob/6ffe4ff2b681143fd38565fd3e416b1a42bf6f2d/src/spaces/jolt_physics_direct_space_state_3d.cpp#L676-L775)._

But how would i go about this more generally? 
Lets say i would like to check out another Method with all its individual steps. 
Where could i look up what is Godot Call is connected to what Godot Jolt Call?

**Reply by mihe** [reply]
> In none of those examples i am able to get a [KinematicCollision3D.GetCollisionCount()](https://docs.godotengine.org/en/stable/classes/class_kinematiccollision3d.html#class-kinematiccollision3d-method-get-collision-count) result of more than 1.

As you can see in [the documentation](https://docs.godotengine.org/en/stable/classes/class_physicsbody3d.html#class-physicsbody3d-method-move-and-collide) of `move_and_collide` you need to override the default value for the `max_collisions` argument if you want more than 1 collision per call.

> Where could i look up what is Godot Call is connected to what Godot Jolt Call?

Generally you would look up the source code for the corresponding scene node in Godot's codebase, like `PhysicsBody3D`, `RigidBody3D`, etc. and then look out for some kind of `PhysicsServer3D` call, which will map pretty directly to any of the methods found [here](https://github.com/godot-jolt/godot-jolt/blob/master/src/servers/jolt_physics_server_3d.cpp). However, some things by pass the physics server entirely, such as `PhysicsDirectBodyState3D` and `PhysicsDirectSpaceState3D`, the implementations for which you'll find [here](https://github.com/godot-jolt/godot-jolt/blob/master/src/objects/jolt_physics_direct_body_state_3d.cpp) and [here](https://github.com/godot-jolt/godot-jolt/blob/master/src/spaces/jolt_physics_direct_space_state_3d.cpp) respectively.
